tbot_contrib/gpio.py

```diff
@@ -0,0 +1,56 @@
+import tbot
+from tbot.machine import linux
+
+
+class Gpio:
+    def __init__(self, host: linux.LinuxShell, gpio_number: int):
+
+        self.host = host
+        self._gpio_number = str(gpio_number)
+        self._gpio_sysclass_path = self.host.fsroot / "sys/class/gpio"
+        self._gpio_path = self._gpio_sysclass_path / f"gpio{self._gpio_number}"
+        self._export()
+
+    def _export(self) -> None:
+        if not self._gpio_path.is_dir():
+            self.host.exec0(
+                "echo",
+                str(self._gpio_number),
+                linux.RedirStdout(self._gpio_sysclass_path / "export"),
+            )
+
+    def _is_in_direction(self) -> bool:
+        direction = self.host.exec0("cat", self._gpio_path / "direction")
+        return direction.strip().lower() == "in"
+
+    def set_in_direction(self) -> None:
+        if self._is_in_direction():
+            pass
+        else:
+            self.host.exec0(
+                "echo", "in", linux.RedirStdout(self._gpio_path / "direction")
+            )
+
+    def set_out_direction(self) -> None:
+        if not self._is_in_direction():
+            pass
+        else:
+            self.host.exec0(
+                "echo", "out", linux.RedirStdout(self._gpio_path / "direction")
+            )
+
+    def get_direction(self) -> str:
+        return self.host.exec0("cat", self._gpio_path / "direction").strip()
+
+    def set_value(self, value: str) -> None:
+        if self._is_in_direction():
+            tbot.log.message("You can not set value from a GPIO that has out direction")
+        else:
+            self.host.exec0("echo", value, linux.Raw(">"), self._gpio_path / "value")
+
+    def get_value(self) -> str:
+        if self._is_in_direction():
+            return self.host.exec0("cat", self._gpio_path / "value").strip()
+        else:
+            tbot.log.message("You can not get value from a GPIO that has in direction")
+            return ""
```

Comment:
gpio could read a wrong value if you read from an output set gpio, I'm going to create cache value so if it is set to output value the get_value() will return the last output value that was set

Reply:
> gpio could read a wrong value if you read from an output set gpio

Is this really true?  The [kernel documentation](https://www.kernel.org/doc/html/latest/driver-api/gpio/legacy.html#paths-in-sysfs) does not really have anything to say regarding this:

> **“value”** ... reads as either 0 (low) or 1 (high). If the GPIO is configured as an output, this value may be written; any nonzero value is treated as high.

---

I think we should read the value from the file in any case; this would for example allow a read-back test checking that `value` does indeed read the last written value.

Reply:
Kernel documentation does not say a lot because it is SOC specific. On some SOC is possible, on most of them not. Kernel calls the internal GPIO driver and of course it reads from register, but it is not ensured that the value read is the value of the signal on the pin. On i.MX, this depends on the pinmux if the SION bit is set or not. If value is read from pin, we could have misunderstanding. And in any case, if a GPIO is output and we set it, the value is what we set. I also suggest that value should be set together with direction in case of output, exactly as this is done in kernel / libgpiod.